src/EditorFeatures/Core.Wpf/InlineRename/AbstractEditorInlineRenameService.SymbolRenameInfo.cs

```diff
@@ -241,6 +238,30 @@ public bool TryOnAfterGlobalSymbolRenamed(Workspace workspace, IEnumerable<Docum
                 return _refactorNotifyServices.TryOnAfterGlobalSymbolRenamed(workspace, changedDocumentIDs, RenameSymbol,
                     this.GetFinalSymbolName(replacementText), throwOnFailure: false);
             }
+
+            public InlineRenameFileRenameInfo GetFileRenameInfo()
+            {
+                var kindSupportsRename = RenameSymbol.Kind == SymbolKind.NamedType;
+                var allowRename = kindSupportsRename 
+                    && (RenameSymbol.Locations.Length == 1)
+                    && (OriginalNameMatches(_document, RenameSymbol.Name) || ContainsOnlyOneType(_document));
+
+                return new InlineRenameFileRenameInfo(kindSupportsRename, allowRename);
+
+                // Local Functions
+
+                static bool OriginalNameMatches(Document document, string name)
+                {
+                    return document.Name.Substring(0, document.Name.Length - 3) == name;
+                }
+
+                static bool ContainsOnlyOneType(Document document)
+                {
+                    var syntaxRoot = document.GetSyntaxRootSynchronously(CancellationToken.None);
+                    var syntaxService = document.GetLanguageService<ISyntaxFactsService>();
+                    return syntaxRoot.DescendantNodesAndSelf().Where(n => syntaxService.IsTypeDeclaration(n)).Count() == 1;
```

Comment:
this is pretty costly.  you *every* node of the tree.  def try this out on a large file.  i'd recommend finding the largest roslyn files (both codegen'ed and not) to see the impact of this.  If bad, you might need to be smarter here and not dive deeper than necessary.

Reply:
This may be an argument for just checking that the type name already matches the file name. We can always add this functionality in later if needed, but for now avoid it due to perf concerns.